geoipupdate supports environment variables for config options

ugexe opened the pull request:
Previously any configuration of `geoipupdate` had to happen through a config file and command line arguments. This updates `geoipupdate` to understand configuration through environment variables, which can also override values set through the config file (but not those set by command line arguments).

ugexe committed:
Refactor config code to better accommodate multiple config sources

The current config code mixes constructing values, setting defaults,
validation, etc. which makes it difficult to reason about how a
configuration is constructed. This refactors the config package
to more clearly separate the various steps of config construction,
as well as providing a way for e.g. a Proxy url from one config
to work with the ProxyUserPassword from an overriding config (which
itself does not provide a Proxy url). It also makes it easier to
add additional configuration fields and/or normalize those fields
since those steps now come after applying any overrides.

ugexe committed:
Add environment variable configuration options

Previous various parts of the config could only be set in a config
file (and some parts by the config file or command line options).
This updates the config package to also set config options based
on environment variables after the config file and before command
line options are used.

ugexe committed:
Use environment variables in docker entry instead of config file

ugexe committed:
Update docs to reference enviroment variable configuration

ugexe commented on dev-bin/make-man-pages.pl:
```diff
@@ -19,6 +19,12 @@ sub main {
     );
+    _make_man(
+        'GeoIP.env',
+        9,
```
I just picked a number 😬

horgh replied:
Is 9 even a thing? I think there's conventions around this. 5 probably is fine.

ugexe commented on doc/GeoIP.env.md:
I'm not sure if having this as a separate document is the best way to do this

horgh replied:
Yeah, we could probably stick with the docs in GeoIP.conf I think. This seems duplicative. I think we could probably not have GeoIP.env in the repo too. Is that commonly done?

ugexe replied:
I don't think its too common. I had originally included it because I thought maybe I would use it with Docker's `env_file`, but that wasn't the case.

ugexe commented on pkg/geoipupdate/config.go:
```diff
@@ -120,10 +118,67 @@ func NewConfig( //nolint: gocyclo // long but breaking it up may be worse
+	// Reset values that were only needed to communicate information between
+	// config overrides.
+
+	config.proxyURL = ""
```
I'm not sure if we want to do this or not. It does make testing a little easier at least.

ugexe commented on pkg/geoipupdate/config.go:
```diff
@@ -39,6 +39,10 @@ type Config struct {
 	// Proxy is host name or IP address of a proxy server.
 	Proxy *url.URL
+	// proxyURL is the host value of Proxy
+	proxyURL string
```
I was trying to find a better way to handle the situation where the proxy url is set by the config and the proxy user password is set by an environment variable, but failed. The problem being that `Proxy` is a `*url.URL` composed of both of those values.

ugexe commented on pkg/geoipupdate/config.go:
```diff
@@ -98,20 +102,14 @@ func WithOutput(val bool) Option {
+// finally by flag overrides provided by flagOptions. Values from the later
+// override the former.
+func NewConfig(
 	path string,
```
I think it would be nice if we could make `path` not required since users may provide their entire configuration without using a config file

ugexe replied:
If we want to go ahead with a major version release I could add this change to this PR

horgh replied:
Yeah, I'm not sure. Perhaps we should go ahead and do that. It is kind of odd how things are right now.

horgh commented on doc/GeoIP.conf.md:
```diff
@@ -17,66 +17,79 @@ sensitive.
 `LicenseKey`
 
-:   Your case-sensitive MaxMind license key.
+:   Your case-sensitive MaxMind license key. This can be overriden at run time
```
"overridden"

horgh commented on doc/GeoIP.conf.md:
```diff
@@ -17,66 +17,79 @@ sensitive.
 
-:   Your case-sensitive MaxMind license key.
+:   Your case-sensitive MaxMind license key. This can be overriden at run time
+    by the `GEOIPUPDATE_LICENSE_KEY` environment variables.
```
"variable"

horgh commented on doc/GeoIP.conf.md:
```diff
@@ -17,66 +17,79 @@ sensitive.
 :	The maximum number of parallel database downloads. The default is
 	1, which means that databases will be downloaded sequentially. This can be
-	overriden at runtime by the `--parallelism` command line argument.
+	overriden at runtime by the `GEOIPUPDATE_PARALLELISM` environment variable
```
"overridden"

horgh commented on doc/GeoIP.conf.md:
```diff
@@ -17,66 +17,79 @@ sensitive.
 :   The proxy user name and password, separated by a colon. For instance,
-    `username:password`.
+    `username:password`. This can be overridden at run time by the
+    `GEOIPUPDATE_USER_PASSWORD` environment variable.
```
How about `GEOIPUPDATE_PROXY_USER_PASSWORD`?

horgh commented on docker/entry.sh:
```diff
@@ -42,33 +42,11 @@ if [ -z "$GEOIPUPDATE_ACCOUNT_ID" ] || [ -z  "$GEOIPUPDATE_LICENSE_KEY" ] || [ -
+# Create an empty configuration file. All configuration is provided via
+# environment variables or command line options, but geoipupdate still
+# expects a configuration file to exist.
 echo "# STATE: Creating configuration file at $conf_file"
```
Maybe we could drop this line then. It might just be confusing.

We should update docker.md as well regarding the config file.

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -98,20 +102,14 @@ func WithOutput(val bool) Option {
-// file.
-func NewConfig( //nolint: gocyclo // long but breaking it up may be worse
+// NewConfig create a new configuration and populates it based on a config
+// file point to by 'path', then by various environment variables, and then
```
"pointed"

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -39,6 +39,10 @@ type Config struct {
 	Proxy *url.URL
+	// proxyURL is the host value of Proxy
+	proxyURL string
+	// proxyUserainfo is the userinfo value of Proxy
```
"proxyUserInfo"

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -166,79 +221,154 @@ func NewConfig( //nolint: gocyclo // long but breaking it up may be worse
+			return errors.New("`PreserveFileTimes' must be 0 or 1")
+		}
+		if value == "1" {
+			config.PreserveFileTimes = true
```
How about `config.PreserveFileTimes = value == "1"`?

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -166,79 +221,154 @@ func NewConfig( //nolint: gocyclo // long but breaking it up may be worse
+			return errors.New("'Verbose' must be 0 or 1")
+		}
+		if value == "1" {
+			config.Verbose = true
```
Similar to the above

horgh commented on pkg/geoipupdate/config_test.go:
```diff
@@ -425,6 +429,137 @@ EditionIDs    GeoLite2-City      GeoLite2-Country
+		},
+	}
+
+	tempFh, err := os.CreateTemp("", "conf-test")
```
Could we use t.TempDir()?

horgh commented:
We should also update the changelog

ugexe committed:
Remove environment config specific documentation

The various environment variables are referenced in the config
file document already, so we can remove these as mostly duplicative.

ugexe committed:
Fix typos

ugexe committed:
Remove confusing message

While it is true we are still creating a configuration file, we
aren't actually writing anything to it anymore and thus anyone
who reads that message probably wouldn't be interested in it any
longer.

ugexe committed:
Simplify some boolean config assignment

ugexe committed:
Use t.TempDir() instead of os.CreateTemp()

ugexe committed:
Mention environment config options in changelog

ugexe committed:
Update expected value in documentation

ugexe committed:
Support $GEOIPUPDATE_CONF_FILE in geoipupdate

ugexe committed:
Update error to reflect non-file config errors

ugexe committed:
Use GEOIPUPDATE_DB_DIR instead of --database-directory

ugexe commented on cmd/geoipupdate/args.go:
```diff
@@ -19,10 +19,15 @@ type Args struct {
 }
 
 func getArgs() *Args {
+	confFileDefault := vars.DefaultConfigFile
```
This logic is here instead of `pkg/config` because we can't tell if the path passed to `config.go NewConfig(path, ...)` came from the command line (and thus we shouldn't override it)

horgh commented on CHANGELOG.md:
```diff
@@ -1,5 +1,26 @@
+    * `GEOIPUPDATE_PROXY`
+    * `GEOIPUPDATE_PROXY_USER_PASSWORD`
+    * `GEOIPUPDATE_RETRY_FOR`
+    * `GEOIPUPDATE_VERBOSE`
```
Can we mention this one in the docs somewhere? Not just in the docker docs that is. Maybe in the `geoipupdate` doc?

horgh commented on doc/GeoIP.conf.md:
```diff
@@ -17,66 +17,79 @@ sensitive.
 :	The maximum number of parallel database downloads. The default is
 	1, which means that databases will be downloaded sequentially. This can be
-	overriden at runtime by the `--parallelism` command line argument.
+	overridden at runtime by the `GEOIPUPDATE_PARALLELISM` environment variable
```
Pre-existing, but this should be "run time" probably, like the others.

horgh commented on doc/docker.md:
```diff
@@ -33,9 +33,10 @@ The following are optional:
-  `geoipupdate` takes. Set to **anything** (e.g., `1`) to enable.
-* `GEOIPUPDATE_CONF_FILE` - The path where the configuration file will be
-  written. The default is `/etc/GeoIP.conf`.
+  `geoipupdate` takes. Set to `1` to enable.
```
Hmm, so requiring it be `1` will be a behaviour change. Perhaps we should make it work as originally described, unless we are going to do a major release. We should mention it in the change log if we're changing it.

horgh commented on doc/docker.md:
```diff
@@ -33,9 +33,10 @@ The following are optional:
-  written. The default is `/etc/GeoIP.conf`.
+  `geoipupdate` takes. Set to `1` to enable.
+* `GEOIPUPDATE_CONF_FILE` - The path of a configuration file to be used by
+  `geoipupdate`. If this file does not exist, an empty file will be created.
```
Saying an empty file will be written seems odd, especially with no explanation. Maybe we can say something about it being optional. Actually the behaviour is going to be different now given the config file can be read now rather than written. We should clarify how this works and say something about it in the changelog I think. We could potentially just not take `GEOIPUPDATE_CONF_FILE` as an env var generally and remove it from this doc.

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -98,20 +102,14 @@ func WithOutput(val bool) Option {
-// file.
-func NewConfig( //nolint: gocyclo // long but breaking it up may be worse
+// NewConfig create a new configuration and populates it based on a config
+// file poinedt to by 'path', then by various environment variables, and then
```
"pointed"

ugexe committed:
Make config file path optional

Previously a configuration file was required, even if all required
configuration was supplied through other means. This removes that
requirement, making the configuration file optional.

ugexe committed:
Use original truth criteria when setting Verbose

The original verbose option behavior enabled it if any value was
used, and would not error if a value besides 0 or 1 was used. This
returns to the original behavior of allowing any value to be used.

ugexe committed:
Fix typos and spacing

ugexe commented on pkg/geoipupdate/config.go:
```diff
@@ -118,12 +132,21 @@ func NewConfig(
-	err := setConfigFromFile(config, path)
+	// Potentially populate config.configFilePath. We will rerun this function
+	// again later to ensure the flag values override env variables.
+	err := setConfigFromFlags(config, flagOptions...)
```
Admittedly this is a bit strange: we apply the flag options here at the start, and again a few blocks lower. I wasn't sure how else to allow a new optional flag to be passed in that itself influenced behavior that we normally do before processing those flags. I'm not sure if this is the ideal way to handle this.

ugexe committed:
Add related environment variables in CLI docs

ugexe committed:
Mention breaking changes in changelog

horgh commented on CHANGELOG.md:
```diff
@@ -1,5 +1,31 @@
+    * `GEOIPUPDATE_DB_DIR`
+    * `GEOIPUPDATE_EDITION_IDS`
+    * `GEOIPUPDATE_HOST`
+    * `GEOIPUPDATE_LICENSE_KEY`
```
I'm wondering now if we should use the `GEOIPUPDATE_LICENSE_KEY_FILE` (and account ID one) that the docker image uses rather than support these. I'm worried it will encourage a practice of putting secrets in the environment, which I believe is not great. Sorry for keeping bringing up new things 😬. Maybe we could bring it up in slack and see what others think.

ugexe replied:
Yeah it is hard to say. I agree with the idea, though I didn't think much about it here as I've used something like the following in some CI workflows more than once:
```
environment:
    GITHUB_ACCESS_TOKEN:
        secure: EPqbi0m6WbH/4m84G1mHPyJ9jf9J90jFw23raWJC1Js62Owlmww0fmexoVEp
```

horgh replied:
Yeah, it does seem like a lot of tools take secrets directly like that. Maybe it's ok. Internally we're trying not to do it as a best practice I believe, and since this docker image currently avoids it I'm leaning to thinking we should continue avoiding it.

ugexe replied:
I added support for the `_FILE` variants, but have left in the e.g. `GEOIPUPDATE_ACCOUNT_ID` env vars for now based on https://maxmind.slack.com/archives/C8D64BH6D/p1688565557869959

horgh commented on cmd/geoipupdate/args.go:
```diff
@@ -19,10 +19,15 @@ type Args struct {
 }
 
 func getArgs() *Args {
+	confFileDefault := vars.DefaultConfigFile
```
If I want to configure based on env vars rather than a config file, will that work? I'm thinking that there will always be a non-empty config file specified given the default so we will always try to use it. In the case where I'm wanting to use env vars only, would that error? Not sure how to solve this if so. I guess we'd have to set `GEOIPUPDATE_CONF_FILE=""` maybe?

We also should think about how docker users will work given they may have been passing that env var already.

ugexe replied:
Yeah, `GEOIPUPDATE_CONFIG_FILE=""` would have to be set to avoid reading the config file altogether. To avoid using any of the values of that config file one would have to supply each option as an environment variable or command line flag -- the file would still get read but the values should be overridden.

The only other thing I can think of is convoluted and doesn't really solve it ideally either. Something like create the config using flags first, then environment variables, and then the config file. Each time it would only update fields that haven't yet been updated. When it reaches the time to set config from the config file, it could check if the current config has any fields that are zero values and if not avoid reading the file altogether. But that would mean every optional value would need to be set explicitly, so I don't think thats what we want to do.

horgh commented on doc/geoipupdate.md:
```diff
@@ -72,9 +75,10 @@ runs `geoipupdate` on each Wednesday at noon:
-variable.
+To use with a proxy server, update your `GeoIP.conf` file as specified in
+the `GeoIP.conf` man page or set the `GEOIPUPDATE_PROXY` environment variable
+as specified in the `GeoIP.env` man page. Alternatively, set the `http_proxy`
```
We shouldn't reference `GeoIP.env` now.

horgh commented on pkg/geoipupdate/config.go:
```diff
@@ -19,6 +19,9 @@ import (
 	AccountID int
+	// confFile is the path to any configuration file used when
+	// potentially populating Config fields.
+	configFile *string
```
Why a pointer?

ugexe committed:
Use non-pointer

ugexe committed:
Add _ACCOUNT_ID_FILE and _LICENSE_KEY_FILE env vars

ugexe committed:
Remove leftover reference to non-existant GeoIP.env

horgh commented on cmd/geoipupdate/main.go:
```diff
@@ -41,14 +41,14 @@ func main() {
 	)
 	if err != nil {
-		fatalLogger(fmt.Sprintf("error loading configuration file %s", args.ConfigFile), err)
+		fatalLogger(fmt.Sprintf("error loading configuration %s", args.ConfigFile), err)
```
Maybe we should not have `args.ConfigFile` in this message now?

ugexe committed:
Make configuration error message more generic

Loading a configuration can fail in ways that do not involve a
config file. This removes part of an error message that implies
otherwise.

ugexe commented:
If we consider the account ID a secret, I wonder if we should change the error message
```
$ GEOIPUPDATE_CONF_FILE="" GEOIPUPDATE_ACCOUNT_ID="1 " ./geoipupdate
error loading configuration: invalid account ID format: strconv.Atoi: parsing "1 ": invalid syntax
```

horgh commented:
Yeah, making that message more generic seems fine. Up to you if you want to add it to this PR or not though :-P

horgh commented on doc/docker.md:
```diff
@@ -10,16 +10,22 @@ The source code is available on [GitHub](https://github.com/maxmind/geoipupdate)
 
-* `GEOIPUPDATE_ACCOUNT_ID` - Your MaxMind account ID.
-* `GEOIPUPDATE_LICENSE_KEY` - Your case-sensitive MaxMind license key.
+
```
Stray blank line I think?

horgh commented on doc/docker.md:
```diff
@@ -10,16 +10,22 @@ The source code is available on [GitHub](https://github.com/maxmind/geoipupdate)
   `GeoIP2-City` would download the GeoIP2 City database (`GeoIP2-City`).
 
+One of:
+* `GEOIPUPDATE_ACCOUNT_ID` - Your MaxMind account ID.
```
Should we have blank lines above this and the start of the next list? Not sure how the docker markdown renderer will show them. I assume it's fine either way though 🤷

ugexe replied:
Yeah I'm not sure either. It rendered fine in the VSCode markdown editor, but I changed it just in case.

ugexe committed:
Avoid leaking account id in error message

ugexe committed:
Tidy whitespace, fix typo

horgh commented on doc/docker.md:
```diff
@@ -10,16 +10,23 @@ The source code is available on [GitHub](https://github.com/maxmind/geoipupdate)
+One of:
+
+* `GEOIPUPDATE_ACCOUNT_ID` - Your MaxMind account ID.
+* `GEOIPUPDATE_ACCOUNT_IDFILE` - A file containing your MaxMind account ID.
```
`ID_FILE`?

oschwald commented on docker/entry.sh:
```diff
@@ -14,67 +14,36 @@ term_handler() {
 while true; do
     echo "# STATE: Running geoipupdate"
-    /usr/bin/geoipupdate -d "$database_dir" -f "$conf_file" $flags 1>$log_file
+    /usr/bin/geoipupdate $flags 1>$log_file
```
This didn't work as apparently we still need to provide a `GeoIP.conf` file.

oschwald commented on docker/entry.sh:
```diff
@@ -14,67 +14,36 @@ term_handler() {
-if ! [ -z "$GEOIPUPDATE_CONF_FILE" ]; then
-  conf_file=$GEOIPUPDATE_CONF_FILE
+if [ -z "$GEOIPUPDATE_DB_DIR" ]; then
+  GEOIPUPDATE_DB_DIR="$database_dir"
```
Doesn't this need to be exported?
